--- a/zsys-setup	2020-12-09 18:23:51.494450002 +0100
+++ b/zsys-setup	2020-12-09 18:46:09.438848739 +0100
@@ -500,7 +549,13 @@
 		swapoff "${SWAPFILE}"
 	fi
 	# Convert to MiB to align the size on the size of a block
-	SWAPVOLSIZE=$(( SWAPSIZE / 1024 / 1024 ))
+	# for suspend to disk we need a swap partition as large as the mounted RAM
+	# SWAPVOLSIZE=$(( SWAPSIZE / 1024 / 1024 ))
+	# get RAM size
+	SWAPVOLSIZE=$(cat /proc/meminfo | grep MemTotal: | sed -n 's|^[^:]*:\W*\([0-9]*\).*|\1|p')
+	echo "Test: >$SWAPVOLSIZE<" > /home/xubuntu/test.txt
+	# do alignent to GB boarder
+	SWAPVOLSIZE=$(( (($SWAPVOLSIZE / 1024) + 1023) & 0xffffc00 ))
 
 	prepare_target "${TARGET}"
 	format_disk "${DISK}" "${PARTBASE}" "${PARTESP}" "${PARTBPOOL}" "${PARTRPOOL}" "${SWAPVOLSIZE}"
